src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.ValueTaskSource.cs

```diff
@@ -29,42 +29,26 @@ private unsafe class ValueTaskSource : IValueTaskSource<int>, IValueTaskSource
             private bool _cancellationHasBeenRegistered;
 #endif
 
-            public static ValueTaskSource Create(
-                AsyncWindowsFileStreamStrategy strategy,
-                PreAllocatedOverlapped? preallocatedOverlapped,
-                ReadOnlyMemory<byte> memory)
-            {
-                // If the memory passed in is the strategy's internal buffer, we can use the base AwaitableProvider,
-                // which has a PreAllocatedOverlapped with the memory already pinned.  Otherwise, we use the derived
-                // MemoryAwaitableProvider, which Retains the memory, which will result in less pinning in the case
-                // where the underlying memory is backed by pre-pinned buffers.
-                return preallocatedOverlapped != null &&
-                       MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> buffer) &&
-                       preallocatedOverlapped.IsUserObject(buffer.Array) ?
-                            new ValueTaskSource(strategy, preallocatedOverlapped, buffer.Array) :
-                            new MemoryValueTaskSource(strategy, memory);
-            }
-
-            protected ValueTaskSource(
-                AsyncWindowsFileStreamStrategy strategy,
-                PreAllocatedOverlapped? preallocatedOverlapped,
-                byte[]? bytes)
+            internal ValueTaskSource(AsyncWindowsFileStreamStrategy strategy)
             {
                 _strategy = strategy;
-                _result = TaskSourceCodes.NoResult;
+                _preallocatedOverlapped = new PreAllocatedOverlapped(s_ioCallback, this, null);
 
                 _source = default;
                 _source.RunContinuationsAsynchronously = true;
+            }
 
-                _overlapped = bytes != null &&
-                              _strategy.CompareExchangeCurrentOverlappedOwner(this, null) == null ?
-                              _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(preallocatedOverlapped!) : // allocated when buffer was created, and buffer is non-null
-                              _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(s_ioCallback, this, bytes);
+            internal NativeOverlapped* Configure(ReadOnlyMemory<byte> memory)
+            {
+                _result = TaskSourceCodes.NoResult;
+                _source.Reset();
```

Comment:
It'd be better to do this in the TryToReuse method (just before storing it into the field) rather than here.  Primary reason for that is we don't want its fields storing references to things unnecessarily while its cached in the FileStream.